lib/saluki-components/src/sources/checks/mod.rs

```diff
@@ -75,27 +83,48 @@ impl ChecksConfiguration {
         self.autodiscovery_provider = Some(Arc::new(provider));
         self
     }
+
+    /// Sets the hostname provider to use.
+    pub fn with_hostname_provider<A>(mut self, provider: A) -> Self
+    where
+        A: HostProvider<Error = GenericError> + Send + Sync + 'static,
+    {
+        self.hostname_provider = Some(Arc::new(provider));
+        self
+    }
 }
 
 #[async_trait]
 impl SourceBuilder for ChecksConfiguration {
     async fn build(&self, _context: ComponentContext) -> Result<Box<dyn Source + Send>, GenericError> {
         match &self.autodiscovery_provider {
-            Some(autodiscovery) => Ok(Box::new(ChecksSource {
-                autodiscovery: Arc::clone(autodiscovery),
-                check_runners: self.check_runners,
-                custom_checks_dirs: if !self.additional_checksd.is_empty() {
-                    Some(self.additional_checksd.split(",").map(|s| s.to_string()).collect())
-                } else {
-                    None
-                },
-            })),
+            Some(autodiscovery) => match &self.hostname_provider {
+                Some(host) => Ok(Box::new(ChecksSource {
+                    autodiscovery: Arc::clone(autodiscovery),
+                    check_runners: self.check_runners,
+                    custom_checks_dirs: if !self.additional_checksd.is_empty() {
+                        Some(self.additional_checksd.split(",").map(|s| s.to_string()).collect())
+                    } else {
+                        None
+                    },
+                    configuration: self.full_configuration.clone(),
+                    host: Arc::clone(host),
```

Comment:
You could just query the hostname here and pass it directly instead of having to pass through a `HostProvider` impl all the way to `ChecksSource`.